envoy/api/v2/route/route.proto

```diff
@@ -539,6 +539,21 @@ message RedirectAction {
   string host_redirect = 1;
 
   // The path portion of the URL will be swapped with this value.
+  // [#comment:TODO(bmadhavan)]: Add support for dynamic redirect msg]
+  // A dynamic redirect msg can be created by specifying a %<match_string>% (usually a prefix)
```

Comment:
Maybe it's just me but I still find this behavior kind of confusing to grok. @zuercher thoughts?

Reply:
@mattklein123 Actually I have the same feeling. I can think of issues with extensibility if we decide to go with just %<match_string>% as we might not be able to specify cases like only *path* and not *query* string.
I am almost there with the implementation where I use `path_redirect: "%PATH_AND_QUERY_AFTER('hello')%"` or  `path_redirect: "%PATH_AFTER('hello')%"` . I should be able to send the envoy PR and a corresponding data-plane-api PR by tomorrow which should clarify most of our questions. Thanks.

Reply:
This is a tad bit confusing. Especially when the prefix match involves longer paths. This also slightly feels a bit unstructured.
Here is a different take on this whole thing:

What you essentially want is 
1. path/prefix match
2. some rewrite on the *matched* portion (prefix_rewrite)
3. Route/redirect

This should satisfy the "dynamic" redirect thing you have. You could repurpose the path redirect code to become a rewrite internally, and tweak the code such that path redirect or host redirect always takes the final mutated path/host and does its business.

The static+dynamic combo you have is awfully close to regex match with variables. Its like

```
match=~/(find)(.*?)
rewrite: /foo/bar/$1/...
redirect to the rewritten path
```

The above format is lot more generic than what you have here, the only disadvantage being using regex. We don't have this variable thing support (AFAIK, the last I saw that code). But adding that would probably be more easier than the option you propose here, IMO.

What do you think ? @mattklein123 / @bmadhavan

Reply:
I definitely think introducing function-like %-expressions here will make things worse. What's the purpose of specifying `%PATH_AFTER('hello')%`? Do you want to match /hello/foo and then have the redirect path contain part of the matched path prefix?

I was under the impression from the original ticket that this was meant to be path prefix matching and the ability to emit the remainder of the path at some point within a path redirect. I was expecting an example config like

`match: { prefix: "/find" } redirect: { path_redirect: "/v2.0/find/%PATH_SUFFIX%" }`

Which would do a prefix rewrite. This is effectively @rshriram's example without regular expressions. 

I fully expect someone will want a regex-matched route to be redirected to a path generated from capture groups in the regex before long. This feature could be written in terms of that, at the expense of performing regex matches. The reason for this to stand as a separate feature is that the prefix matching should be faster and perhaps easier to read and therefore reason about.

Reply:
Thanks @rshriram/@zeurcher for the inputs. 
Yes. The combination of path/prefix match + prefix_rewrite + redirect is what I am looking for. 
Add the way I thought of implementing this is similair to what you described i.e. repurpose the path redirect code to become a rewrite internally and use it

@zeurcher *What's the purpose of specifying %PATH_AFTER('hello')%? Do you want to match /hello/foo and then have the redirect path contain part of the matched path prefix?*
Yes. This is very similair to a prefix_rewite for a redirect where part of a prefix is re-used. 
%PATH_AFTER('hello')% can handle `match: { prefix: /hello/foo }` will match request /hello/foo/bar/red/ and redirect it to /foo/bar/red/.
While %PATH_SUFFIX% can only do  `match: { prefix: /hello/foo }` will match request /hello/foo/bar/red/ and redirect it to /bar/red/.

So, if I understood right, use a `rewrite` field under redirect and use it with prefix/path match just like prefix_rewrite. (replacing the matched prefix/path with the value in the field) 
For regex: I am still not very clear on how the regex variable approach will work for all dynamic+static use case. Will I be able to add new static content before and after the regex variables ?

Two open questions,
* Currently with v2 API we have `prefix_rewrite` only under route & cannot be configured with redirect ? Are you suggesting to add a new `rewrite` field under "redirect": "{...}" as well ? 
* ALso I would want to redirect with just the URI path from the request & strip off the query string. Will it be ok to expose a new field like `retain_query: false` under redirect
to acheive this ?

Reply:
You have to tweak the current logic to make redirect work with rewrite. And for everything else, you probably need a regex_rewrite or something that will do the job. A regex that that strips off the query shouldn't be that expensive would it?

Reply:
> Yes. This is very similair to a prefix_rewite for a redirect where part of a prefix is re-used.
> %PATH_AFTER('hello')% can handle match: { prefix: /hello/foo } will match request /hello/foo/bar/red/ and redirect it to /foo/bar/red/.
> While %PATH_SUFFIX% can only do match: { prefix: /hello/foo } will match request /hello/foo/bar/red/ and redirect it to /bar/red/.

Can you just mach a prefix of `/hello/foo` and rewrite it as `/foo/%PATH_SUFFIX%` to get what you want? (But see below.)

> So, if I understood right, use a rewrite field under redirect and use it with prefix/path match just like prefix_rewrite. (replacing the matched prefix/path with the value in the field)

Yes, this would actually eliminate the need for %PATH_SUFFIX% altogether. I think this answers your first open question.

> For regex: I am still not very clear on how the regex variable approach will work for all dynamic+static use case. Will I be able to add new static content before and after the regex variables ?

Not clear that this should go into prefix_rewrite, but the idea is that you could configure:

`match: { regex: "/hello/(foo|bar)/(.*)" } redirect: { prefix_rewrite: "/$1/qux/$2" }`

to map /hello/foo/xyz to /foo/qux/xyz or some such. I think we'd want to implement this in in RouteAction as well. Unless you really need support for this, I'd leave it out for now.

> ALso I would want to redirect with just the URI path from the request & strip off the query string. Will it be ok to expose a new field like retain_query: false under redirect to acheive this ?

I think that's ok, but @mattklein123 should chime in. You should reverse the sense of the flag though -- `strip_query: true` would enable this and the field's default value would be false.

Reply:
@rshriram Even though I agree with using `prefix_rewrite`, I just have one concern with using the `prefix_rewrite` under route section to work with redirect section. Not only will the json validator fail as both route and redirect cannot be configured together, i feel it will be counter intuitive for someone to configure under route and use it for redirect. How about adding a similar field under *redirect:{...}* ? your thoughts ???

Reply:
Thanks @zuercher . That answers all my questions. 
1) I will eliminate %PATH_SUFFIX% approach completely and just make `prefix_rewrite` work with `path_redirect`. It will be great if @mattklein123 can confirm if we want to reuse the field under `route` or create a new one under `redirect`. I would prefer the latter.
2) Once we come up with the regex approach to rewrite we can decide if we will reuse the prefix_rewrite or come up with a new field. I don't need that right now and do not prefer working on it now. If needed I will work on it later. 
3) I agree on `strip_query: false` as an option and default as `false` ( to maintain current envoy behavior ).  Will wait for @mattklein123 response about adding this field. 
Once clarified, I will update the PR with the new ones. Thanks guys !!

Reply:
Hey thanks for sorting through this. 

1. Agreed let's add a new field under RedirectAction. We can add regex support later if needed.
3. Agreed.